File: cplusplus/tests/models-test.cpp
#include "fixtures/helper.hpp"
#include <gtest/gtest.h>
#include <shapes/base/models.hpp>

using namespace std;

namespace helper = shapes::tests::fixtures;

namespace shapes::tests {
constexpr double DELTA = 0.001;

namespace circle_tests {
    TEST(CircleTest, TestCircleArea) {
        auto circle = helper::setupCircle();
        const double result = circle.area();
        const double expected = 754.767;
        EXPECT_TRUE(abs(result - expected) < DELTA);
    }

    TEST(CircleTest, TestCirclePerimeter) {
        auto circle = helper::setupCircle();
        const double result = circle.perimeter();
        const double expected = 97.389;
        EXPECT_TRUE(abs(result - expected) < DELTA);
    }
} // namespace circle_tests

namespace rectangle_tests {
    TEST(RectangleTest, TestRectangleArea) {
        auto rectangle = helper::setupRectangle();
        const double result = rectangle.area();
        const double expected = 756.437;
        EXPECT_TRUE(abs(result - expected) < DELTA);
    }

    TEST(RectangleTest, TestRectanglePerimeter) {
        auto rectangle = helper::setupRectangle();
        const double result = rectangle.perimeter();
        const double expected = 112.0;
        EXPECT_TRUE(abs(result - expected) < DELTA);
    }
} // namespace rectangle_tests

namespace square_tests {
    TEST(SquareTest, TestSquareArea) {
        auto square = helper::setupSquare();
        const double result = square.area();
        const double expected = 105.062;
        EXPECT_TRUE(abs(result - expected) < DELTA);
    }

    TEST(SquareTest, TestSquarePerimeter) {
        auto square = helper::setupSquare();
        const double result = square.perimeter();
        const double expected = 41.0;
        EXPECT_TRUE(abs(result - expected) < DELTA);
    }
} // namespace square_tests
} // namespace shapes::tests

File: rust/tests/models_test.rs
mod fixtures {
    pub mod helper;
}

use fixtures::helper;
use shapes::base::commons::Dimensional;

const DELTA: f64 = 0.001;

mod circle_tests {

    use super::*;

    #[test]
    fn test_circle_area() {
        let circle = helper::setup_circle();
        let result = circle.area();
        let expected = 754.767;
        assert!((result - expected).abs() < DELTA);
    }

    #[test]
    fn test_circle_circumference() {
        let circle = helper::setup_circle();
        let result = circle.perimeter();
        let expected = 97.389;
        assert!((result - expected).abs() < DELTA);
    }
}

mod rectangle_tests {
    use super::*;

    #[test]
    fn test_rectangle_area() {
        let rectangle = helper::setup_rectangle();
        let result = rectangle.area();
        let expected = 756.437;
        assert!((result - expected).abs() < DELTA);
    }

    #[test]
    fn test_rectangle_perimeter() {
        let rectangle = helper::setup_rectangle();
        let result = rectangle.perimeter();
        let expected = 112.0;
        assert!((result - expected).abs() < DELTA);
    }
}

mod square_tests {
    use super::*;

    #[test]
    fn test_square_area() {
        let square = helper::setup_square();
        let result = square.area();
        let expected = 105.062;
        assert!((result - expected).abs() < DELTA);
    }

    #[test]
    fn test_square_perimeter() {
        let square = helper::setup_square();
        let result = square.perimeter();
        let expected = 41.0;
        assert!((result - expected).abs() < DELTA);
    }
}
